lib/widgets/compose_box.dart

```diff
@@ -0,0 +1,252 @@
+import 'dart:convert';
+import 'package:flutter/material.dart';
+import 'package:zulip/interactions/showErrorDialog.dart';
+
+import 'app.dart';
+import '../api/route/messages.dart';
+
+enum TopicValidationError { mandatoryButEmpty, tooLong }
+
+class TopicTextEditingController extends TextEditingController {
+  // Mimic TextEditingController's non-default constructors, which are not
+  // inherited. (Is there a more transparent way to do this?)
+  TopicTextEditingController({ String? text }) : super(text: text);
+  TopicTextEditingController.fromValue(TextEditingValue? value) : super.fromValue(value);
+
+  // TODO: subscribe to this value:
+  //   https://zulip.com/help/require-topics
+  final mandatory = true;
+
+  String get textNormalized {
+    String trimmed = text.trim();
+    return trimmed.isEmpty ? kNoTopicTopic : trimmed;
+  }
+
+  Set<TopicValidationError> get validationErrors {
+    Set<TopicValidationError> result = Set.identity();
+    if (mandatory && textNormalized == kNoTopicTopic) {
+      result.add(TopicValidationError.mandatoryButEmpty);
+    }
+    if (textNormalized.length > kMaxTopicLength) {
+      result.add(TopicValidationError.tooLong);
+    }
+    return result;
+  }
+}
+
+enum ContentValidationError {
+  empty,
+  tooLong,
+  // TODO: upload in progress
+  // TODO: quote-and-reply in progress
+}
+
+class ContentTextEditingController extends TextEditingController {
+  // Mimic TextEditingController's non-default constructors, which are not
+  // inherited. (Is there a more transparent way to do this?)
+  ContentTextEditingController({ String? text }) : super(text: text);
+  ContentTextEditingController.fromValue(TextEditingValue? value) : super.fromValue(value);
+
+  String get textNormalized {
+    return text.trim();
+  }
+
+  Set<ContentValidationError> get validationErrors {
+    Set<ContentValidationError> result = Set.identity();
+    if (textNormalized.isEmpty) {
+      result.add(ContentValidationError.empty);
+    }
+    if (utf8.encode(textNormalized).length > kMaxMessageLengthBytes) {
```

Comment:
Eep, this sounds slow to do on every build: converting the whole text from one encoding to another.

Particularly as we're rebuilding on every keystroke, given those listeners. (That should probably change too — this compose-box widget has a very large build tree, most of it doesn't care about the values, and we should avoid rebuilding the parts that don't — but we'll still need to re-ask this question on every keystroke, because it's part of how to show the send button.)

One optimization that would solve it for common cases is to use the fact that the number of UTF-8 bytes is at most 3 times the number of UTF-16 code units, i.e. at most `3 * textNormalized.length`. (That's because a three-byte UTF-8 sequence holds 16 bits: 4 in the first byte, 6 in each trailing byte. So every codepoint that fits in 16 bits — i.e., every codepoint that fits into a single UTF-16 code unit — fits into three bytes of UTF-8. And no codepoint takes more than four bytes of UTF-8.)

That'd make it fast up to 3333 UTF-16 code units. But then there'd be a cliff where when the message gets long, things suddenly _really_ get slow.

In preference to that cliff, I think I'd be happier with either of:
* When you get up to `kMaxMessageLengthBytes/3` code units, that's it — that's the limit. If you were writing mostly ASCII text, you could have gone on 3x longer before the server would have cared. But you probably didn't *really* want that, did you?

  Especially when writing on a phone.
* When your message gets long, we just assume that you're writing mostly ASCII text (the stuff that gets one UTF-8 byte per character) and that it's fine, up to `kMaxMessageLengthBytes` code units. When you actually try to send, we check then.

OK, and after writing those two options out, I think I definitely prefer the former: just set the limit conservatively.

Reply:
> Particularly as we're rebuilding on every keystroke, given those listeners. (That should probably change too — this compose-box widget has a very large build tree, most of it doesn't care about the values, and we should avoid rebuilding the parts that don't […])

(But this part I'm happy to let be while merging this, and I'll try refactoring it in that direction later.)

Reply:
OK, turns out the server limit here is actually in Unicode codepoints:
https://chat.zulip.org/#narrow/stream/412-api-documentation/topic/max.20message.20size/near/1508001

That's `textNormalized.runes.length` in Dart. But that's the length of an iterator — it works by walking through the whole string to count the codepoints. Not *quite* as expensive as `utf8.encode`, because it doesn't have to allocate a copy of the whole thing, but still not cheap.

So I think what I'd want to do is to just treat the maximum as a max number of UTF-16 code units, i.e. a bound on `textNormalized.length`.

In reality every codepoint occupies either one or two UTF-16 code units. So this means that if you write a giant message entirely made of emoji, we might cut you off at just 5000 emoji while the server would have accepted up to 10000; same for medieval-era Chinese characters; and more realistically if you have a smattering of those then we'll cut you off a little bit early.